pkg/cmd/util/clientcmd/factory.go

```diff
@@ -277,11 +281,18 @@ func (c *clientCache) ClientConfigForVersion(version string) (*kclient.Config, e
 		c.defaultConfig = config
 	}
 	// TODO: have a better config copy method
+	if config, ok := c.configs[version]; ok {
+		return config, nil
+	}
+	// TODO: have a better config copy method
 	config := *c.defaultConfig
-	if len(version) != 0 {
-		config.Version = version
+	negotiatedVersion, err := kclient.NegotiateVersion(nil, &config, version, latest.Versions)
```

Comment:
Thanks for the detail, now I see. I think I've resolved this by setting up a default client based on the loader fed to the cache, which should be configured with origin defaults. Then I feed that client to the negotiator.